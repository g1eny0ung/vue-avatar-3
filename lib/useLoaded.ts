import { ref, watchEffect } from 'vue-demi'
import type { Ref } from 'vue-demi'
import type { ImageOptions } from './types'

function useLoaded(options: Ref<ImageOptions>) {
  const loading = ref(true)
  const loaded = ref(false)

  function loadImage(options: ImageOptions) {
    const { crossorigin, referrerpolicy, src, srcset } = options

    if (!src && !srcset) {
      return
    }

    // Reset states.
    loading.value = true
    loaded.value = false

    let active = true
    const image = new Image()
    image.onload = () => {
      if (!active) {
        return
      }

      loaded.value = true
      loading.value = false
    }
    image.onerror = () => {
      if (!active) {
        return
      }

      loaded.value = false
      loading.value = false
    }

    if (crossorigin) {
      image.crossOrigin = crossorigin
    }
    if (referrerpolicy) {
      image.referrerPolicy = referrerpolicy
    }
    if (src) {
      image.src = src
    }
    if (srcset) {
      image.srcset = srcset
    }

    return () => {
      active = false
    }
  }

  watchEffect((onCleanup) => {
    const cancel = loadImage(options.value)

    onCleanup(() => {
      typeof cancel === 'function' && cancel()
    })
  })

  return { loading, loaded }
}

export default useLoaded
