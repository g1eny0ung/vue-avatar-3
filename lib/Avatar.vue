<script setup lang="ts">
import { computed } from 'vue-demi'
import type { ImageOptions } from './types'
import useLoaded from './useLoaded'
import { stringAvatar, pickTextColorBasedOnBgColorSimple } from './utils'

const props = withDefaults(
  defineProps<{
    srcSet?: ImageOptions['srcset']
    sizes?: ImageOptions['sizes']
    src?: ImageOptions['src']
    alt?: ImageOptions['alt']
    imgProps?: ImageOptions
    name?: string
    bgColor?: string
    size?: string | number
    variant?: 'circular' | 'rounded' | 'square'
    round?: string | number
    fallback?: boolean
  }>(),
  {
    variant: 'circular',
    fallback: true,
  }
)

const useLoadedOptions = computed(() => ({
  crossorigin: props.imgProps?.crossorigin,
  referrerpolicy: props.imgProps?.referrerpolicy,
  src: props.src || props.imgProps?.src,
  srcset: props.srcSet || props.imgProps?.srcset,
}))

const { loading, loaded } = useLoaded(useLoadedOptions)
const showStringAvatar = computed(
  () => !loading.value && !loaded.value && props.fallback && props.name
)
const stringAvatarComputed = computed(() =>
  props.name ? stringAvatar(props.name) : undefined
)

const defaultSize = '40px'
const size = computed(() =>
  props.size
    ? typeof props.size === 'string'
      ? props.size
      : `${props.size}px`
    : defaultSize
)
</script>

<template>
  <div
    :style="{
      position: 'relative',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      flexShrink: 0,
      width: size,
      height: size,
      backgroundColor:
        bgColor ||
        (showStringAvatar ? stringAvatarComputed?.bgColor : undefined),
      color: bgColor
        ? pickTextColorBasedOnBgColorSimple(bgColor)
        : showStringAvatar
        ? stringAvatarComputed?.textColor
        : undefined,
      borderRadius:
        variant === 'circular'
          ? '50%'
          : variant === 'rounded'
          ? round
            ? typeof round === 'string'
              ? round
              : `${round}px`
            : '8px'
          : undefined,
      userSelect: 'none',
    }"
  >
    <template v-if="showStringAvatar">
      {{ stringAvatarComputed && stringAvatarComputed.children }}
    </template>
    <img
      v-if="loaded"
      :srcset="srcSet"
      :sizes="sizes"
      :src="src"
      :alt="alt"
      :style="{
        position: 'absolute',
        width: '100%',
        height: '100%',
        color: 'transparent', // Hide alt text.
      }"
      v-bind="imgProps"
    />
  </div>
</template>
