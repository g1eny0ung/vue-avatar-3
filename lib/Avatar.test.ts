import { describe, it, expect, vi } from 'vitest'
import { ref } from 'vue'
import { render } from '@testing-library/vue'
import useLoaded from './useLoaded'

import Avatar from './Avatar.vue'

vi.mock('./useLoaded')

describe('Avatar', () => {
  it('renders a image correctly', () => {
    vi.mocked(useLoaded).mockImplementationOnce(() => {
      return {
        loading: ref(false),
        loaded: ref(true),
      }
    })
    const { container, getByRole } = render(Avatar, {
      props: {
        src: 'https://fakeimg.pl/300/',
      },
    })

    expect(container).toMatchSnapshot()
    expect(getByRole('img')).toHaveAttribute('src', 'https://fakeimg.pl/300/')
  })

  it('if the image is not loaded, it should render a letter avatar', () => {
    vi.mocked(useLoaded).mockImplementationOnce(() => {
      return {
        loading: ref(false),
        loaded: ref(false),
      }
    })
    const { container, getByText } = render(Avatar, {
      props: {
        src: 'https://fakeimg.pl/300/',
        name: 'John Doe',
      },
    })

    expect(container).toMatchSnapshot()
    expect(getByText('JD')).toBeInTheDocument()
    expect(getByText('JD')).toHaveStyle('color: #fff')
  })

  it('if only the name is provided, it should render a letter avatar', () => {
    vi.mocked(useLoaded).mockImplementationOnce(() => {
      return {
        loading: ref(false),
        loaded: ref(false),
      }
    })
    const { container, getByText } = render(Avatar, {
      props: {
        name: 'John Doe',
      },
    })

    expect(container).toMatchSnapshot()
    expect(getByText('JD')).toBeInTheDocument()
  })

  it('if bgColor is provided, it should render a letter avatar with the provided background color', () => {
    vi.mocked(useLoaded).mockImplementationOnce(() => {
      return {
        loading: ref(false),
        loaded: ref(false),
      }
    })
    const { container, getByText } = render(Avatar, {
      props: {
        name: 'John Doe',
        bgColor: '#ffffff',
      },
    })

    expect(container).toMatchSnapshot()
    expect(getByText('JD')).toHaveStyle('background-color: #ffffff')
    expect(getByText('JD')).toHaveStyle('color: #000')
  })

  it('variant: rounded', () => {
    vi.mocked(useLoaded).mockImplementationOnce(() => {
      return {
        loading: ref(false),
        loaded: ref(true),
      }
    })
    const { container } = render(Avatar, {
      props: {
        src: 'https://fakeimg.pl/300/',
        variant: 'rounded',
      },
    })

    expect(container).toMatchSnapshot()
    expect(container.firstChild).toHaveStyle('border-radius: 8px')
  })

  it('variant: rounded + round 4px', () => {
    vi.mocked(useLoaded).mockImplementationOnce(() => {
      return {
        loading: ref(false),
        loaded: ref(true),
      }
    })
    const { container } = render(Avatar, {
      props: {
        src: 'https://fakeimg.pl/300/',
        variant: 'rounded',
        round: 4,
      },
    })

    expect(container).toMatchSnapshot()
    expect(container.firstChild).toHaveStyle('border-radius: 4px')
  })

  it('variant: square', () => {
    vi.mocked(useLoaded).mockImplementationOnce(() => {
      return {
        loading: ref(false),
        loaded: ref(true),
      }
    })
    const { container } = render(Avatar, {
      props: {
        src: 'https://fakeimg.pl/300/',
        variant: 'square',
      },
    })

    expect(container).toMatchSnapshot()
    expect(container.firstChild).not.toHaveStyle('border-radius: 8px')
  })

  it('if fallback is false, it should not render a letter avatar when the image is not loaded', () => {
    vi.mocked(useLoaded).mockImplementationOnce(() => {
      return {
        loading: ref(false),
        loaded: ref(false),
      }
    })
    const { queryByText } = render(Avatar, {
      props: {
        src: 'https://fakeimg.pl/300/',
        name: 'John Doe',
        fallback: false,
      },
    })

    expect(queryByText('JD')).not.toBeInTheDocument()
  })
})
