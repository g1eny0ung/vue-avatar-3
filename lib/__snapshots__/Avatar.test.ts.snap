// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Avatar > if bgColor is provided, it should render a letter avatar with the provided background color 1`] = `
<div>
  <div
    style="position: relative; display: flex; justify-content: center; align-items: center; flex-shrink: 0; width: 40px; height: 40px; background-color: #ffffff; color: #000; border-radius: 50%; user-select: none;"
  >
    
    JD
    
    <!--v-if-->
  </div>
</div>
`;

exports[`Avatar > if only the name is provided, it should render a letter avatar 1`] = `
<div>
  <div
    style="position: relative; display: flex; justify-content: center; align-items: center; flex-shrink: 0; width: 40px; height: 40px; background-color: #a55c80; color: #fff; border-radius: 50%; user-select: none;"
  >
    
    JD
    
    <!--v-if-->
  </div>
</div>
`;

exports[`Avatar > if the image is not loaded, it should render a letter avatar 1`] = `
<div>
  <div
    style="position: relative; display: flex; justify-content: center; align-items: center; flex-shrink: 0; width: 40px; height: 40px; background-color: #a55c80; color: #fff; border-radius: 50%; user-select: none;"
  >
    
    JD
    
    <!--v-if-->
  </div>
</div>
`;

exports[`Avatar > renders a image correctly 1`] = `
<div>
  <div
    style="position: relative; display: flex; justify-content: center; align-items: center; flex-shrink: 0; width: 40px; height: 40px; border-radius: 50%; user-select: none;"
  >
    <!--v-if-->
    <img
      src="https://fakeimg.pl/300/"
      style="position: absolute; width: 100%; height: 100%; color: transparent;"
    />
  </div>
</div>
`;

exports[`Avatar > variant: rounded + round 4px 1`] = `
<div>
  <div
    style="position: relative; display: flex; justify-content: center; align-items: center; flex-shrink: 0; width: 40px; height: 40px; border-radius: 4px; user-select: none;"
  >
    <!--v-if-->
    <img
      src="https://fakeimg.pl/300/"
      style="position: absolute; width: 100%; height: 100%; color: transparent;"
    />
  </div>
</div>
`;

exports[`Avatar > variant: rounded 1`] = `
<div>
  <div
    style="position: relative; display: flex; justify-content: center; align-items: center; flex-shrink: 0; width: 40px; height: 40px; border-radius: 8px; user-select: none;"
  >
    <!--v-if-->
    <img
      src="https://fakeimg.pl/300/"
      style="position: absolute; width: 100%; height: 100%; color: transparent;"
    />
  </div>
</div>
`;

exports[`Avatar > variant: square 1`] = `
<div>
  <div
    style="position: relative; display: flex; justify-content: center; align-items: center; flex-shrink: 0; width: 40px; height: 40px; user-select: none;"
  >
    <!--v-if-->
    <img
      src="https://fakeimg.pl/300/"
      style="position: absolute; width: 100%; height: 100%; color: transparent;"
    />
  </div>
</div>
`;
